druid-shell/src/mouse.rs

```diff
@@ -18,50 +18,240 @@ use crate::kurbo::Point;
+#[derive(PartialEq, Eq, Clone, Copy, Default)]
+pub struct MouseButtons(u8);
+
+impl MouseButtons {
+    /// Create a new empty set.
+    #[inline]
+    pub fn new() -> MouseButtons {
+        MouseButtons(0)
+    }
+
+    /// Add the `button` to the set.
+    #[inline]
+    pub fn add(&mut self, button: MouseButton) {
+        self.0 |= 1 << button as u8;
+    }
+
+    /// Remove the `button` from the set.
+    #[inline]
+    pub fn remove(&mut self, button: MouseButton) {
+        self.0 &= !(1 << button as u8);
+    }
+
+    /// Builder-style method for adding the `button` to the set.
+    #[inline]
+    pub fn with(mut self, button: MouseButton) -> MouseButtons {
+        self.0 |= 1 << button as u8;
+        self
+    }
+
+    /// Builder-style method for removing the `button` from the set.
+    #[inline]
+    pub fn without(mut self, button: MouseButton) -> MouseButtons {
+        self.0 &= !(1 << button as u8);
+        self
+    }
+
+    /// Returns `true` if the `button` is in the set.
+    #[inline]
+    pub fn has(self, button: MouseButton) -> bool {
+        (self.0 & (1 << button as u8)) != 0
+    }
+
+    /// Returns `true` if any button is in the set.
+    #[inline]
+    pub fn has_any(self) -> bool {
+        self.0 != 0
+    }
+
+    /// Returns `true` if the set is empty.
+    #[inline]
+    pub fn has_none(self) -> bool {
+        self.0 == 0
+    }
+
+    /// Returns `true` if [`MouseButton::Left`] is in the set.
+    ///
+    /// [`MouseButton::Left`]: enum.MouseButton.html#variant.Left
+    #[inline]
+    pub fn has_left(self) -> bool {
+        self.has(MouseButton::Left)
+    }
+
+    /// Returns `true` if [`MouseButton::Right`] is in the set.
+    ///
+    /// [`MouseButton::Right`]: enum.MouseButton.html#variant.Right
+    #[inline]
+    pub fn has_right(self) -> bool {
+        self.has(MouseButton::Right)
+    }
+
+    /// Returns `true` if [`MouseButton::Middle`] is in the set.
+    ///
+    /// [`MouseButton::Middle`]: enum.MouseButton.html#variant.Middle
+    #[inline]
+    pub fn has_middle(self) -> bool {
+        self.has(MouseButton::Middle)
+    }
+
+    /// Returns `true` if [`MouseButton::X1`] is in the set.
+    ///
+    /// [`MouseButton::X1`]: enum.MouseButton.html#variant.X1
+    #[inline]
+    pub fn has_x1(self) -> bool {
+        self.has(MouseButton::X1)
+    }
+
+    /// Returns `true` if [`MouseButton::X2`] is in the set.
+    ///
+    /// [`MouseButton::X2`]: enum.MouseButton.html#variant.X2
+    #[inline]
+    pub fn has_x2(self) -> bool {
+        self.has(MouseButton::X2)
+    }
+
+    /// Returns `true` if [`MouseButton::Other`] is in the set.
+    ///
+    /// [`MouseButton::Other`]: enum.MouseButton.html#variant.Other
+    #[inline]
+    pub fn has_other(self) -> bool {
+        self.has(MouseButton::Other)
+    }
+
+    /// Adds all the [`MouseButton`]s in `other` to the set.
+    ///
+    /// [`MouseButton`]: enum.MouseButton.html
+    pub fn union(&mut self, other: MouseButtons) {
```

Comment:
I would expect this to return a new `MouseButtons`

Reply:
You shouldn't, because builder methods start with `with`.

Reply:
this wouldn't be a builder method, it would just be a method that happens to return the union of two sets; similar to the `union` method on collection types like `HashSet`. (These actually return an iterator over the union, but since our type is `Copy` I wouldn't worry about that).

If what you want to do is add all the items from one set to the other, the only `std` api for this would be the [`Extend`](http://doc.rust-lang.org/1.42.0/std/iter/trait.Extend.html#) trait.

Anyway not a blocker or anything, just something that jumps out at me.